### PR TITLE
IRS Init page

### PR DESCRIPTION
**Pull request by Externoak**
Fixes https://github.com/flybywiresim/a32nx/issues/610 
Fixes https://github.com/flybywiresim/a32nx/issues/401

**Summary of Changes**
Added IRS Init page to MCDU.
Button will appear only after a From/To airports have been inserted.
Menu will display Reference coordinates based on origin airport, GPS will appear aligning once at least one IRS knob is on.
When all 3 IRS knobs are on it will display a confirmation message to start aligning the IRS (source https://github.com/flybywiresim/a32nx/issues/401)
Reference coordinates can be modified.
Correct coordinates format is used in IRS INIT page and INIT page. https://github.com/flybywiresim/a32nx/pull/677#issuecomment-692872320

**Screenshots (if necessary)**

image
image

IRS INIT page IRS OFF

image

One IRS ON

image

Three IRS ON

image

Reference coordinates are now modifiable and reference name is correctly set when coordinates match origin airport [source](https://youtu.be/o3--ALDKLUc?t=78). When back to original coordinates origin airport name should appear again. Modified LAT and LONG with up and down keys are shown:
Original coordinates:
image
Modified:
image

You can select your reference coordinates and align on them:
Once you are happy with the coordinantes click ALIGN ON REF:
image
Confirm:
image
It will align on the given reference:
image
Once aligned it shall show ALIGNED ON (GPS or REF)
image

**Additional context**

~~LAT and LONG button works as expected but can't adjust reference LAT and LON at the moment.~~ **Implemented**
~~ALIGN ON REF is currently not implemented.~~ **Implemented**

Discord username (if different from GitHub): Externo#2359

### CODE REVIEW

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:24]: Those variables can be `const` if they're not reassigned ;)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:16]: minor: variables should start with a lowercase letter to differentiate them easily

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:66]:
```suggestion
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:98]:
```suggestion
            lon = false
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:102]:
```suggestion
            lon = true
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:110]:
```suggestion
            if (!SimVar.GetSimVarValue("A32NX_ADIRS_PFD_ALIGNED_FIRST", "Bool") && !SimVar.GetSimVarValue("L:A320_Neo_ADIRS_IN_ALIGN", "Bool")) {
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:120]: `async`` does nothing here since the UI is entirely single-threaded.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:137]: Please fix the spacing in this function, it's hard to read ;)

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js:65]: This is badly indented and hard to understand. Please fix the indentation ;) Also, never use `var`.

Please also remove `console.log`s from release code.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js:94]:
```suggestion
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js
Comment (Benjozork) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js:116]:
```suggestion
        mcdu.onRightInput[2] = () => {
```

**Review by Benjozork — COMMENTED** [review]
Thanks for your PR !

A few things to fix in the code, but looks good so far.

**Comment by CoffeePanda0** [comment]
I'm not sure why, but this branch seems to break the A32NX for me. All screens including radio and a few lights don't work after downloading this.
https://streamable.com/w3tra4 Here is a video showing the broken aircraft

**Comment by Snapmatics** [comment]
Your forgot to do the JSON thingy. 
image

**Comment by CoffeePanda0** [comment]
> 
> 
> Your forgot to do the JSON thingy.
> image

Yeah just built it again, apologies I forgot to build

**Comment by Benjozork** [comment]
> 
> 
> > Your forgot to do the JSON thingy.
> > image
> 
> Yeah just built it again, apologies I forgot to build
> Worth building before the merge for sure though

Nah layout.json is built automatically by CI, no need for it to be done in the PR.

**Comment by Snapmatics** [comment]
> > > Your forgot to do the JSON thingy.
> > > image
> > 
> > 
> > Yeah just built it again, apologies I forgot to build
> > Worth building before the merge for sure though
> 
> Nah layout.json is built automatically by CI, no need for it to be done in the PR.

Well it was broken for me so I just built it again and that fixed it.

**Comment by Benjozork** [comment]
> Well it was broken for me so I just built it again and that fixed it.

I know. I'm saying that it doesn't need to be done for merge :) It still needs to be done on your part.

**Comment by CoffeePanda0** [comment]
Sorry about that layout.json
After some testing seems to work as intended! Small thing is for some reason I cant change the integ lighting on this branch but I imagine that's just because it is an older branch and should be fine once merged!
Thanks for the work,, @Externoak !

**Comment by Externoak** [comment]
@Benjozork Thanks for the point-outs, I'll do a cleanup of the code.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (AdenFlorian) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:120]: async isn't about multithreading. It means that the function returns a promise. This is fine.

**Comment by AdenFlorian** [comment]
at this point, I think it's fine to push the layout and manifest to make things easier for QA.

**Comment by armankir** [comment]
Seemed to work out as described, but didn't go really in-depth.

**Comment by kiwi2021** [comment]
Please take a look at the conflict so we can merge this.

**Comment by Externoak** [comment]
@kiwi2021 The conflict is due to the layout.json file in the master. It gets regenerated in master branch for every new commit so it will always creates a conflict. I commited it as asked in: https://github.com/flybywiresim/a32nx/pull/677#issuecomment-692346123 , when it's ready to be merged I will ensure no conflicts occur :)

**Comment by DoToAdventures** [comment]
Hello,
I tested this. The functions described work as expected.
I've also tested what happens if you turn off the IRS knobs after you already pressed Confirm Align and it prompts you to do that again once you turn all three on again, which is correct according to Tarek 👍 

One thing I have noticed - The Latitude and Longtitude indicators on the Init page do not appear in the correct format, which he also confirmed. 
QsSPL00
The format seen inside the IRS Init page is correct. 
zA7Iw5b

**Comment by Externoak** [comment]
One thing I'm currently not happy with is that IRS ALIGN IN  X message on ECAM appears (when all 3 Knobs of IRS are active ) before you confirm the alignment.
I will work on it as soon as I have time.

**Comment by Externoak** [comment]
MDCU page now displays LAT and LONG correctly as indicated by https://github.com/flybywiresim/a32nx/pull/677#issuecomment-692789286:

image

ECAM Displays message to align IRS:

image

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (davidwalschots) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:125]: I believe this is equivalent and reduces some complexity that isn't needed:
```suggestion
        autoRefresh();
        function autoRefresh() {
            setTimeout(() => {
                if (mcdu.getTitle() == 'IRS INIT') {
                    CDUIRSInit.ShowPage(mcdu, lon = lon);
                }
            }, 2000);
        }
```

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (davidwalschots) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:37]: **thought**: There's a lot of inconsistency in usage of `;` at the end of lines.
**suggestion**: Consistently add `;` at the end of lines, where applicable.

Comment (Externoak) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js]: Added missing ; to code.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js
Comment (Externoak) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_IRSInit.js:124]: I agree, has been changed accordingly.

**Comment by Externoak** [comment]
Had a little time to implement missing functions.
New functions have been implemented, re-testing should occur to ensure all is up to the standard.

**Comment by starvingjavi** [comment]
Conflict with new update. All screens are black when trying to use it in the new update

**Comment by Externoak** [comment]
@starvingjavi New update is not merged into this branch yet. It will be merge once stable, you can test in the meantime. Black screens are due to layout.json not updated please run build.py see https://github.com/flybywiresim/a32nx/pull/677#issuecomment-692312592

**Comment by starvingjavi** [comment]
What is described works as intended, however I think some of the functionality should be reworked to make it be similar to a study level A320. I was looking at various videos of how to start from Cold & Dark and a lot of the functionality is there. However, I don't know if I am supposed to get a precise Lat and Long as soon as I turn the knobs, since they should be aligning

**Comment by Externoak** [comment]
We could fake the IRS into taking more time to align but at the moment it aligns instantly once activate. 
I'm unsure if this feature is desired right now or when the IRS system gets reworked.

**Comment by davidwalschots** [comment]
> 
> 
> We could fake the IRS into taking more time to align but at the moment it aligns instantly once activated.
> I'm unsure if this feature is desired right now or when the IRS system gets reworked.

From a software design point of view, it makes much more sense to make the IRS system itself responsible for "alignment time". I'd not add it to the MCDU code.

**Comment by Benjozork** [comment]
@starvingjavi @davidwalschots, consider having this functionality in `A32NX_Core`. As @davidwalschots says, having logic like that in a view controller is not good practice.

**Comment by Snapmatics** [comment]
> We could fake the IRS into taking more time to align but at the moment it aligns instantly once activated.
> I'm unsure if this feature is desired right now or when the IRS system gets reworked.

What do you mean fake the IRS?
(From my POV I can tell you that: We can't implement things that could make things easier for use. Our goal is to reach to a study-level aircraft at some point). It would be better to hold-off and get this thing done without rushing.)

If you mean fake with a different meaning then I am sorry, haha.

**Comment by Externoak** [comment]
> > We could fake the IRS into taking more time to align but at the moment it aligns instantly once activated.
> > I'm unsure if this feature is desired right now or when the IRS system gets reworked.
> 
> What do you mean fake the IRS?
> (From my POV I can tell you that: We can't implement things that could make things easier for use. Our goal is to reach to a study-level aircraft at some point). It would be better to hold-off and get this thing done without rushing.)
> 
> If you mean fake with a different meaning then I am sorry, haha.

Faking the IRS was just an idea, although personally I did not like it as @davidonthereg said. At the time it appeared as a more realistic response from the aircraft.

Want I meant to fake was the time the IRS systems take to start nothing else. I've talked about this with @UncleClapton and he said that the IRS systems was being develop at the moment so hopefully there will be no need to fake it and they should work as intended which is great :)

**Comment by camilo625** [comment]
Name: Camilo625 #5575
Date of testing: Current date 23/09/2020
Version of the sim: 1.8.3.0
PR Tested: #677 

Testing technique: Spawn C&D and test IRS Init page while IRS is in alignment 
Steps you've made to test the Pull Request:
1. Spawn SAZB C&D
2. Load SAZB/SADP into INIT
3. Turn IRSs to ON
4. Check IRS Init page

Overall Rating: Good
Conclusions of the testing: Aside from ALIGN ON REF not being implemented, and the instantaneous Lat/Long, the rest seems to work as described.

**Comment by Ares1378** [comment]
Some changes: 

This is how it should look if everything is aligned:
image

**Comment by Externoak** [comment]
@Ares1378 On it.

**Comment by Externoak** [comment]
Correct display when all IRS are aligned. 
image

**Comment by Benjozork** [comment]
Hello, please take note that we have started including a CHANGELOG.md file that requires every PR to add a change-log line that describes its changes. As your PR is not merged yet, we kindly ask that you add this information following the included directives in the file.

Disclaimer: This is a copy-and-pasted message.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js
Comment (lhoenig) [inline A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js:298]: If I understand correctly, this is the same functionality as the standard function String.prototype.padStart(), used in a lot of code already.

**Comment by Externoak** [comment]
PR has been updated and tested with new files from 1.9.3 and method padStart is now used as indicated by @lhoenig.

File: A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js
Comment (Externoak) [reply A32NX/html_ui/Pages/VCockpit/Instruments/Airliners/A320_Neo/CDU/A320_Neo_CDU_InitPage.js]: Correct, modified accordingly

**Comment by AdenFlorian** [comment]
What references did you base this on? I believe the CONFIRM ALIGN only shows when you press ALIGN ON REF:
http://www.a320dp.com/A320_DP/nav-flight-management/sys-14.6.8A.html

And also, I don't think you need to confirm anything to align on GPS, as it automatically does that by default:
https://www.youtube.com/watch?v=o3--ALDKLUc

@Ares1378 can you confirm?

**Review by AdenFlorian — CHANGES_REQUESTED** [review]
requesting changes to the functionality.

**Comment by Externoak** [comment]
Waiting for confirmation from @Ares1378 https://github.com/flybywiresim/a32nx/pull/677#issuecomment-701588915

File: .github/CHANGELOG.md
Comment (NathanInnes) [inline .github/CHANGELOG.md:19]: Change to current month

**Comment by AdenFlorian** [comment]
@Externoak FYI you don't have to wait for Ares. I recommend asking around on the discord.

**Comment by Ares1378** [comment]
@AdenFlorian that is the incorrect MCDU on a320dp.com that is reflecting the THALES MCDU.
@Externoak These are taken from the honeywell manual that can be found in the discord if you do not have it you can message me personally on discord Ares#3394
image
image
image
image
image

From what I can see everything looks correct visually! Just make sure the code reflects what the manual describes. When this is ready to test let me know and I can test it out.

**Comment by Ares1378** [comment]
> MDCU page now displays LAT and LONG correctly as indicated by [#677 (comment)](https://github.com/flybywiresim/a32nx/pull/677#issuecomment-692789286):
> 
> image
> 
> ECAM Displays message to align IRS:
> 
> image

I cannot find the ECAM: MCDU CONFIRM IRS ALIGN the only ecam I can find is NAV IR NOT ALIGNED when there is an actual problem with the alignment. Is there anyone else that can comment on this that is an IRL a320
 pilot I am unable to find this in a NEO QRH
Also the LAT and LONG is not displayed on the Honeywell MCDU INIT A page please refer to PR #880

File: .github/CHANGELOG.md
Comment (Externoak) [reply .github/CHANGELOG.md]: Resolved

**Review by AdenFlorian — APPROVED** [review]
reviewed and tested! LGTM!

**Pull request merged**
